File: keskiarvo_vs_mediaani/Python/keskiarvo_vs_mediaani.py
# Kokeillaan mediaanin ja keskiarvon eroa. Kuvissa (30kpl) on oppilaita
# satunnaisissa kohdissa, ja kamera oli jalustalla luokkahuoneessa. Otetaan
# toisaalta keskiarvot ja toisaalta mediaanit pikseliarvoista.
# Lopputulokset ovat hyvin erilaiset!
#
# Samuli Siltanen huhtikuu 2021
# Matlab -> Python Ville Tilvis kesäkuu 2021

import numpy as np
import matplotlib.pyplot as plt


# Kuvien lukumäärä
Nim = 30

# Alustetaan matriisit, joihin tallennetaam keskiarvot ja mediaanit
im_ave = np.zeros([2000,2997,3])
im_median = np.zeros([2000,2997,3])
im_4D = np.zeros([2000,2997,3,Nim])

print("Ladataan kuvat:")

# Avataan kuvat yksi kerrallaan
for iii in range (0,Nim):
    fname = '../_kuvat/IMGP'+str(1423+iii)+'.jpg'
    im_orig = plt.imread(fname,'jpg');
    
    # Lisätään tämänhetkinen kuva pakkaan
    im_4D[:,:,:,iii] = im_orig;
    
    # Seuraa ajoa
    print(iii+1,"/",Nim)

print("Lasketaan keskiarvo ja mediaani...")

im_ave = np.mean(im_4D,axis=3)/255;
im_median = np.median(im_4D,axis=3)/255;

print("Valmis!")
print("")
print("Näytetään kuvat...")

# Vähennetään keskiarvokuva ja
# mediaanikuva tyhjän kuvan
# punaisesta värikanavasta
im0 = np.array(plt.imread('../_kuvat/IMGP1444.jpg','jpg'))/255

error1 = np.abs(im_ave-im0)
error2 = np.abs(im_median-im0)
errorpic = np.concatenate((error1,error2),axis=1)
errorpic = errorpic/np.max(errorpic[:,:,0])
errorpic = np.power(errorpic,0.3)

#Katsotaan kuvia
plt.subplot(2,1,1)  
plt.imshow(np.concatenate((im_ave,im_median),axis=1))
plt.axis('off')
plt.gcf().set_dpi(600)

plt.subplot(2,1,2) 
plt.imshow(errorpic[:,:,0],cmap='gray', interpolation='none')
plt.axis('off')
plt.gcf().set_dpi(600)

plt.show()


print("Valmis!")
print("")
print("Tallennetaan kuvat...")

# Tallennetaan kuvat
plt.imsave('../_kuvat/im_average.jpg',im_ave,);
plt.imsave('../_kuvat/im_median.jpg',im_median);

print("Valmis!")

File: keskiarvo_vs_mediaani/Matlab/keskiarvo_vs_mediaani.m
% Kokeillaan mediaanin ja keskiarvon eroa. Kuvissa (30kpl) on oppilaita
% satunnaisissa kohdissa, ja kamera oli jalustalla luokkahuoneessa. Otetaan
% toisaalta keskiarvot ja toisaalta mediaanit pikseliarvoista.
% Lopputulokset ovat hyvin erilaiset!
%
% Samuli Siltanen huhtikuu 2021

% Kuvien lukumäärä
Nim = 30;

% Alustetaan matriisit, joihin tallennetaam keskiarvot ja mediaanit
im_ave = zeros(2000,2997,3);
im_median = zeros(2000,2997,3);
im_4D = zeros(2000,2997,3,Nim);

% Avataan kuvat yksi kerrallaan
for iii = 1:Nim
    fname = ['../_kuvat/IMGP',num2str(1422+iii),'.jpg'];
    im_orig = imread(fname,'jpg');
    
    % Muutetaan pikseliarvot liukuluvuiksi
    im = double(im_orig);
    
    % Lisätään tämänhetkinen kuva pakkaan
    im_4D(:,:,:,iii) = im;
    
    % Seuraa ajoa
    disp([iii,Nim])
end
im_ave = mean(im_4D,4);
im_median = median(im_4D,4);

% Katsotaan kuvaa
figure(1)
clf
imshow(uint8(round([im_ave,im_median])))
axis equal
colormap gray

% Vähennetään keskiarvokuva ja
% mediaanikuva tyhjän kuvan
% punaisesta värikanavasta
im0 = double(imread('../_kuvat/IMGP1444.jpg','jpg'));

errorpic = [abs(im_ave-im0),abs(im_median-im0)];
errorpic = errorpic/max(errorpic(:));
errorpic = errorpic.^(0.3);
figure(2)
clf
imshow(uint8(round(255*errorpic)))
axis equal

% Tallennetaan kuvat
imwrite(uint8(im_ave),'../_kuvat/im_average.jpg');
imwrite(uint8(im_median),'../_kuvat/im_median.jpg');
